### DISCUSSION TITLE
High-level models? CUSTOM modes?

### DISCUSSION THREAD

**Discussion by 6lj6** [helix-editor/helix#5767] [Ideas] [answered]
In helix, there are plenty of low-level operations like `delete-by-word-to-1D_forward` or `move-cursor-withinSelection-by-2D_line-down`.

Any script or APIs or Plugins in HELIX to define custom mode based on High-level models?

e.g. press "t2" to enter a TOML file mode
1. So I can just press "owner" then ENTER without "[]", it will auto add it
2. then I input "name", then TAB without "=", it will auto add it
3. then I input "Tom Preston-Werner"
    1. then ENTER without "", it will auto add it then jump out [owner]
    2. or TAB wihtout "", it will auto add it then wait me to input next attribute "dob"
    3. Q to go back to TOML_view mode, then I can move around by element of TOML instead of text characters
    4. S to switch the style of VALUE of key/VALUE pair from [] to {}, also in VALUE I can CRUD element of this array in different level of the arrayTree.
    5. and more... according to the spec of TOML or Rust lang, so we can choose code by meaning instead of compose every character of the text. `High-level` VS `Low-level` models.

Any CRUD operations can be defined in particular mode with simple syntax instead use text low-level char-level operations.
```


# This is a TOML document

title = "TOML Example"

[owner]
name = "Tom Preston-Werner"
dob = 1979-05-27T07:32:00-08:00

[database]
enabled = true
ports = [ 8000, 8001, 8002 ]
data = [ ["delta", "phi"], [3.14] ]
temp_targets = { cpu = 79.5, case = 72.0 }

[servers]

[servers.alpha]
ip = "10.0.0.1"
role = "frontend"

[servers.beta]
ip = "10.0.0.2"
role = "backend"


```

**Comment by the-mikedavis** [comment, marked as answer]
There isn't currently a plugin system or APIs for these kinds of changes but this kind of language-specific behavior would most likely be accomplished with a plugin. See #3806
